### PR TITLE
Allow to mention groupfolder users in file chats

### LINKED ISSUE #2139: Users in a group folder can not be mentioned in File rooms
Pending from #2102

## How to test (scenario 1)
- Install the [Group Folders app](https://github.com/nextcloud/groupfolders)
- Go to the _Users_ settings and create a group with several users, for example, _admin_ and _user0_
- Go to the _Group folders_ settings and create a new folder; associate it to the group just created
- Log in as one of the users in the group, for example, admin
- Create or upload a file to the group folder
- Open the details view for the new file
- Open the Chat tab
- Type _@_ in the _New message_ input

## Expected result
The candidate mentions include _user0_

## Actual result
The candidate mentions do not include _user0_

## How to test (scenario 2)
- Install the [Group Folders app](https://github.com/nextcloud/groupfolders)
- Go to the _Users_ settings and create a group with several users, for example, _admin_ and _user0_
- Go to the _Group folders_ settings and create a new folder; associate it to the group just created
- Log in as one of the users in the group, for example, admin
- Create or upload a file to the group folder
- Open the details view for the new file
- Open the Chat tab
- Send a new message with the content _Hello @user0_
- Log in as _user0_
- Open Talk

## Expected result
The room for the file in the group folder is shown in the list of rooms; if the Notifications app is installed a new notification is shown.

## Actual result
The room for the file in the group folder is not shown in the list of rooms (that is, the user was not added to the room); if the Notifications app is installed there is no notification.

### PR DESCRIPTION
**Pull request by nickvergessen**
Fixes #2139

### CODE REVIEW

**Review by danxuliu — CHANGES_REQUESTED** [review]
Tested with group folders and external storages and mentions now work.

However, there is a problem with external storages. After mentioning a user with access to the file through the external storage, but not through a share, the mentioned user will not be able to join the conversation (there will be an error and the conversation will be removed from the list).

Either mentioning other users with access to the file should be limited only to group folders or something similar to https://github.com/nextcloud/spreed/pull/2102 should be done for external storages too.

**Review by danxuliu — CHANGES_REQUESTED** [review]
Integration tests are not happy.

**Comment by nickvergessen** [comment]
Rebased as CI result was purged

**Comment by nickvergessen** [comment]
Rebased once more, lets see

**Comment by nickvergessen** [comment]
Fixed the integration test and rebased

**Comment by nickvergessen** [comment]
Three last failing tests:
```
/drone/server/apps/spreed/tests/integration/features/conversation/files.feature:223
/drone/server/apps/spreed/tests/integration/features/conversation/files.feature:450
/drone/server/apps/spreed/tests/integration/features/conversation/files.feature:476
```

They all 3 assume that the owner of a file can no longer access a room of a file they joined before when there is no share anymore. From my POV that is not the expected behaviour and with default expiration date can cause unintended information loss. I would therefor change the expected result.
Did I misunderstand something? @danxuliu

**Comment by danxuliu** [comment]
Yes, currently if a share is removed the participants can no longer access the associated room.

From my point of view that is the right behaviour, though :-) After all the room is linked to a file, so if the user can no longer access the file I would not expect the user to be able to access the room either.

Having said that, yes, if a share expires that could lead to information loss. Maybe if a share is deleted/expired the room should be innaccessible for any other participant except the owner, just like the file itself? Although if the owner never joined the room that could be a problem... :thinking:

**Comment by nickvergessen** [comment]
> From my point of view that is the right behaviour, though :-) After all the room is linked to a file, so if the user can no longer access the file I would not expect the user to be able to access the room either.

Well the test cases we talk about are from owner side, so they can still access the file.

**Comment by danxuliu** [comment]
> > From my point of view that is the right behaviour, though :-) After all the room is linked to a file, so if the user can no longer access the file I would not expect the user to be able to access the room either.
> 
> Well the test cases we talk about are from owner side, so they can still access the file.

Right, I missed the _owner_ part :facepalm: Fine by me then.

**Comment by nickvergessen** [comment]
Updated the tests

**Review by danxuliu — COMMENTED** [review]
I adjusted again the integration tests, as the tests for non-owner users no longer with access to the file (_participant2_) are already covered in their own scenarios. I also moved again "joins room" after "deletes shares" in the first scenario, as that ensures that the owner will be able to join the room after the share is deleted even if she was not in the conversation before.

Independently of that, I found an issue:
- As an admin, add user A and user B to a group
- As an admin, create an external storage for that group (and allow sharing in the external storage)
- In the regular window, log in as user A
- Open the Files app
- Open the shared folder and create a file
- In a private window, log in as user B (this is needed for the mount cache to be updated)
- In the regular window, reload the folder
- In the Chat tab for the new file, start/join the conversation
- Mention user B
- In the Sharing tab, share the file by link
- In the private window, open Talk and join the conversation
- As an admin, remove user B from the group
- In the private window, reload Talk and try to join the conversation again

The user will be kicked out from the conversation. However, if the shared link is opened user B will be able to join the conversation in the sidebar of the share page, and it will be possible to open the conversation directly from Talk from that point on. If instead of an external storage a group folder is used the same problem appears (although I would swear that I also saw it working with a group folder :shrug:).

This should not be a very common scenario, though, so fine by me to move it to an issue and fix it at a later point.

File: tests/integration/features/conversation/files.feature [hunk 1]
```diff
@@ -449,21 +447,17 @@ Feature: conversation/files
   # Participants are removed from the room for a no longer shared file once they
   # try to join the room again, but not when the file is unshared.
 
-  Scenario: owner is not participant of room for file no longer shared
+  Scenario: owner is participant of room for file no longer shared
     Given user "participant1" shares "welcome.txt" with user "participant2" with OCS 100
     And user "participant2" accepts last share
     And user "participant1" gets the room for path "welcome.txt" with 200 (v1)
     And user "participant1" joins room "file welcome.txt room" with 200 (v4)
-    And user "participant2" joins room "file welcome.txt room" with 200 (v4)
```
Comment (nickvergessen) [inline tests/integration/features/conversation/files.feature]: why did you delete all the participant2 tests?

**Comment by nickvergessen** [comment]
Tests passed, fixing up

**Review by danxuliu — APPROVED** [review]
Tested and works :+1: (except for the issue mentioned before, and another small issue mentioned below)

Besides the mentions, file rooms also work now for files in external storages :tada: 

One detail worth noting (not a blocker, only something to keep in mind): if a user who is not logged in is added to a group that user will not appear in the mentions until the user logs in again. The opposite is true if the user is removed, that user will still appear as a candidate in the mentions until the user logs in again.

I guess this happens due to the use of the user mount cache, which I think that will not get updated until the file system is mounted again for the user.

Similarly, the need for a cache update might cause too some unexpected behaviours for freshly created external storages and group folders, like the storage/folder not being seen as accessible by a second user and thus not being able to start/join a conversation in a file until that second user reloads the page after the storage/folder was created and the file was added to it :shrug:

Probably there is nothing that can be done about that, but I have no idea.

**Comment by nickvergessen** [comment]
That's acceptable and also happens with activities and file comments 🥴

**Pull request merged**